### PR TITLE
🌱 vcsim: implement cluster proxy to allow more e2e tests to be compatible to vcsim

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by chrischdi** [comment]
Enabled vcsim on following tests to see where it fails.

Disclamer: might not make sense for all of these tests to enable it.

```
Cluster creation with anti affined nodes [vcsim]
When upgrading a workload cluster using ClusterClass and testing K8S conformance [vcsim]  [supervisor] [Conformance] [K8s-Upgrade] [ClusterClass]
When upgrading a workload cluster using ClusterClass [vcsim] [supervisor] [ClusterClass]
When testing ClusterClass changes [vcsim] [supervisor] [ClusterClass]
When testing clusterctl upgrades using ClusterClass (CAPV 1.10=>current, CAPI 1.7=>1.7) [vcsim] [ClusterClass]
When testing clusterctl upgrades using ClusterClass (CAPV 1.9=>current, CAPI 1.6=>1.7) [vcsim] [ClusterClass]
ClusterClass Creation using Cluster API quick-start test and IPAM Provider [vcsim] [ClusterClass]
Label nodes with ESXi host info [vcsim]
Cluster creation with storage policy [vcsim]
```

File: test/e2e/e2e_suite_test.go [hunk 1]
```diff
@@ -272,7 +272,11 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 			Finder: vsphereFinder,
 		}))
 	}
-	bootstrapClusterProxy = framework.NewClusterProxy("bootstrap", kubeconfigPath, initScheme(), clusterProxyOptions...)
+	if testTarget == VCSimTestTarget {
+		bootstrapClusterProxy = vspherevcsim.NewClusterProxy("bootstrap", kubeconfigPath, initScheme(), clusterProxyOptions...)
```
Comment (fabriziopandini) [inline test/e2e/e2e_suite_test.go:276]: I will add a comment about why we need this

**Comment by chrischdi** [comment]
/test pull-cluster-api-provider-vsphere-test-main

flake

**Comment by chrischdi** [comment]
Bumped to CAPI main and rebased

/test pull-cluster-api-provider-vsphere-e2e-vcsim-govmomi-main
/test pull-cluster-api-provider-vsphere-e2e-vcsim-supervisor-main

File: test/framework/vcsim/cluster_proxy.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	// Get the port on the vcsim pod from the RESTConfig of the workload cluster.
+	restConfig := p.ClusterProxy.GetWorkloadCluster(ctx, namespace, name).GetRESTConfig()
+	u, err := url.Parse(restConfig.Host)
+	Expect(err).ToNot(HaveOccurred())
+	port, err := strconv.Atoi(u.Port())
+	Expect(err).ToNot(HaveOccurred())
+
+	// Create a dialer which proxies through the kube-apiserver to the vcsim pod's port where the simulated kube-apiserver
+	// is running.
+	proxyDialer, err := inmemoryproxy.NewDialer(inmemoryproxy.Proxy{
```
Comment (sbueringer) [inline test/framework/vcsim/cluster_proxy.go:66]: Really elegant that we can reuse this here. Nice!

**Review by sbueringer — COMMENTED** [review]
Nice one!

**Comment by chrischdi** [comment]
Forgot to run `make generate-modules`

/test pull-cluster-api-provider-vsphere-e2e-vcsim-govmomi-main
/test pull-cluster-api-provider-vsphere-e2e-vcsim-supervisor-main

**Pull request merged**